### PR TITLE
Add log following

### PR DESCRIPTION
**Pull request by timbrown5**
Extension of: #175 

Video:
https://user-images.githubusercontent.com/4447764/225406899-deae7b82-b1e2-4bcb-bc8e-f3717db6772f.mov

Changes:
- Add a new ConsoleLogStream component to house the Virtuoso list and logic.
-- Allow the Virtuoso list to follow logs and request more output (when following).
--- This changes the behaviour a bit so that the Plugin will only request new log output when following the log. If we add a search feature this might need a re-think but it simplifies things.
- Updated onSelect code to expand a newly focused stage and expand the last step card.
- The Console will stick on the selected node. The node can be unselected by clicking it again, at which point the console will follow the log again.

I've been using this Pipeline for testing:
```
pipeline {
    agent any

    stages {
        stage('Hello') {
            steps {
                script {
                    for(int i=1; i<=10; i++) {
                        stage("Stage - ${i}") {
                            sh 'for i in {1..120}; do sleep 1; echo "Slept ${i} times"; done'
                        }
                    }
                }
            }
        }
    }
}
```

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira

### CODE REVIEW

File: src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx [hunk 1]
```diff
@@ -458,4 +472,38 @@ export default class PipelineConsole extends React.Component<
...
+          stepId={stepId}
+          key={`console-log-${stepId}`}
+        />
+      );
+    } else {
+      return <div>Placeholder...</div>;
+    }
+  }
+
+  zzrender() {
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx:502]: is this intentional?

**Review by timja — COMMENTED** [review]
Can you add some padding / margin to the scroll to bottom text please?
----

image

----

Unrelated but noticed while testing, as you get to 3 digit line numbers there's no enough padding it's touching the side:

image

----

Is it worth adding some sort of spinner or loading bar to show that the step is still running / loading text? Currently the only way you know it's still doing something is when new lines appear

**Comment by timbrown5** [comment]
Sorry for the slow reply. I am adding some tests to the Pipeline view,
which required a small refactor to make parts easily mockable.
This has broken the polling so will put up an update once that's fixed.
One the plus side I think I've resolved your comments on my local copy.

**Review by timja — COMMENTED** [review]
3 issues I noticed.

1. certain times overflow and touch the expander, a lot more likely to be noticed while watching the time increase:
image
2. it's not possible to look at other logs while the job is running, it force jumps you to watch the running logs
3. when it force jumps you it doesn't update the active selection on the left panel (see stage 4 active but stage 8 logs)
image

**Comment by timbrown5** [comment]
> 3 issues I noticed.
> 
>     1. certain times overflow and touch the expander, a lot more likely to be noticed while watching the time increase:
> 
> 
> image
> 2. it's not possible to look at other logs while the job is running, it force jumps you to watch the running logs 3. when it force jumps you it doesn't update the active selection on the left panel (see stage 4 active but stage 8 logs)
> 
> image

I'll have a look. The focus/selected node logic might be a little broken.
I've added the idea of a focused stage - the one open vs a selected stage - the one selected by the user. If the code is changing the selected stage that is probably an issue.

**Comment by timbrown5** [comment]
> 2. it's not possible to look at other logs while the job is running, it force jumps you to watch the running logs.
I think this hasn't been resolved (I fixed some issues with the step selection - and added tests - but not the stage selection).

> 3. when it force jumps you it doesn't update the active selection on the left panel (see stage 4 active but stage 8 logs).
I think the solution to this might resolve the behaviour @stuartrowe reported here:
https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/177#issuecomment-1297575955
Seems the workaround is to manually find the DOM element and focus it yourself. My only worry is that this could cause some unwanted behaviour - changing the user's focus when they are looking at something else (although as there are no input boxes this is likely not a big issue?).

I'm considering update the new behaviour to:
1. Select the TreeItem onNodefocus - this should remove the difference between focused and selected items (the focused item will always be selected).
1. Remove the 'focus' highlight - and move it to 'selected'. This means the highlight will remain even if the user clicks outside of the TreeView.
- Together, these should avoid the limitation @stuartrowe reported here (and 3 above):
https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/177#issuecomment-1297575955
3. If a TreeItem is selected then stick to that item.
1. Allow the user to click the selected item again to unselect it (the highlight will be removed). This will start following the Pipeline again on the next update - won't have an effect when the Pipeline is finished.

Depending on what unselecting an item does, we might need to add a custom CSS class to control highlighting.
Thoughts welcome :).

**Comment by timja** [comment]
Sounds reasonable,

> Allow the user to click the selected item again to unselect it (the highlight will be removed). This will start following the Pipeline again on the next update - won't have an effect when the Pipeline is finished.

❤️

**Comment by timbrown5** [comment]
One issue is that `onSelected` is only fired when the selected field is set (not when it's cleared). Might need to add a custom clickable action to implement the above.

**Comment by timbrown5** [comment]
> One issue is that `onSelected` is only fired when the selected field is set (not when it's cleared). Might need to add a custom clickable action to implement the above.

This statement is incorrect. Don't know what was happening to make it only work 1/2 the time. It's working for me now.

**Comment by timja** [comment]
Is this ready for re-review?

**Comment by timbrown5** [comment]
Sorry, had another look and found a couple of things that needed fixing (including your first issue above).

File: src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx [hunk 2]
```diff
@@ -296,46 +398,30 @@ export default class PipelineConsole extends React.Component<
...
+    this.setState((prevState) => {
+      return {
+        ...prevState,
+        stepBuffers: stepBuffersCopy,
+      };
+    });
   }
 
   handleStepToggle(event: React.SyntheticEvent<{}>, nodeId: string): void {
+    console.log(`Node '${nodeId}' toggled.`);
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx:424]: remove?
```suggestion
```

**Review by timja — APPROVED** [review]
Only thing I noticed when testing is times can still quite easily get cut-off.
image

That's a Mac 13" at full width so a very standard size.

Nonblocking and can be improved later.

----

LGTM

**Pull request merged**